### ISSUE TITLE
[RayCluster] Using the new Suspending / Suspended conditions in RayCluster to ensure the atomicity of the suspend operation

### ISSUE THREAD

**Issue by kevin85421** [kubernetes-sigs/kueue#3199]
@rueian introduced two conditions, `Suspending` and `Suspended`, in the RayCluster CRD in [this PR](https://github.com/ray-project/kuberay/pull/2403), and the field `.Status.State = rayv1.Suspended` will be deprecated in the long term.

With the new conditions, if `spec.suspend` is set to true, KubeRay will first set the `Suspending` condition to true and then begin deleting Pods. If the user sets `spec.suspend` back to false before all Pods are deleted, KubeRay will still wait for all Pods to be deleted, set the `Suspended` condition to true and `Suspending` to false, and then set `Suspended` to false and recreate the Pods.

Without these two conditions, the behavior of setting `spec.suspend` back to false during the suspension process is undefined.

If Kueue maintainers don't have enough bandwidth, @rueian is happy to help.

**Comment by kevin85421** [comment]
cc @andrewsykim @alculquicondor @astefanutti

**Comment by rueian** [comment]
Thank you @kevin85421 for mentioning me. I am happy to help!

**Comment by alculquicondor** [comment]
This is great. It can help us know if there are still running Pods after we have set the suspend field.

cc @mimowo 

We are pretty packed, so we would appreciate the help in the kueue repo to use these conditions.

**Comment by kevin85421** [comment]
Hi @alculquicondor, would you mind providing some guidance on the compatibility between Kueue and KubeRay? For example, which KubeRay versions will be supported by the next Kueue release?

**Comment by alculquicondor** [comment]
We are happy to support the latest released version. However, older versions currently supported should ideally remain supported.

We use the v1 API right now, so hopefully the first version to introduce v1 should remain supported.

IIUC, this can be done, by simply defaulting to the existing implementation if none of the new conditions exist.

**Comment by kevin85421** [comment]
Yeah, it is not hard

**Comment by tenzen-y** [comment]
@kevin85421 @rueian Do you know which version KubeRay started to support the new "Suspending" condition?

**Comment by rueian** [comment]
Hi @tenzen-y, I think it will be released in the next version. v1.2.3 I hope.

**Comment by tenzen-y** [comment]
> Hi @tenzen-y, I think it will be released in the next version. v1.2.3 I hope.

Thanks for letting me know. So, the KubeRay will release the new condition in the next patch version.
If you can contribute to Kueue, I would appreciate it.

**Comment by kevin85421** [comment]
I think the next release should be v1.3.0.

**Comment by tenzen-y** [comment]
> I think the next release should be v1.3.0.

@kevin85421 Thank you for letting us know. Do you have candidates for the v1.3.0 like this month or next month?

**Comment by kevin85421** [comment]
> Do you have candidates for the v1.3.0 like this month or next month?

Maybe next month.

**Comment by tenzen-y** [comment]
> > Do you have candidates for the v1.3.0 like this month or next month?
> 
> Maybe next month.

Thanks, in that case, we can support this Suspending condition after the next Kueue version (0.10.0).

**Comment by tenzen-y** [comment]
@kevin85421 Do you have any plan to stop serving deprecated `.Status.State = rayv1.Suspended` state handling mechanism?

**Comment by tenzen-y** [comment]
> [@tenzen-y](https://github.com/tenzen-y) maybe next major release, but we don't have a clear timeline for the release v2.0.0.

Thank you for letting me know. In that case, we might want to keep the current suspending mechanism (`.status.state` = `suspend`) in the Kueue side to keep supporting older RayJobs. After Ray has been released v2.0, we can use the new way.

cc: @mimowo @mszadkow

**Comment by kevin85421** [comment]
@tenzen-y, does it make sense to support both `Status.State` and `conditions` in Kueue for now? That way, when KubeRay removes the field later, the impact on users will be minimized.

**Comment by mszadkow** [comment]
Isn't that conflicting with `Kueue` being the one to manipulate `spec.suspend`?
Or you talk only about the effects of setting it to false/true by the Kueue.
Because RayCluster has the feature to suspend the cluster, but it cause me some troubles when I wanted to incorporate this in our e2e tests.

**Comment by tenzen-y** [comment]
> Isn't that conflicting with `Kueue` being the one to manipulate `spec.suspend`? Or you talk only about the effects of setting it to false/true by the Kueue. Because RayCluster has the feature to suspend the cluster, but it cause me some troubles when I wanted to incorporate this in our e2e tests.

Good point. Indeed, we do not depend on `.status.Suspend` field in https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/raycluster/raycluster_controller.go

@kevin85421 Could you clarify the expected replacement?

**Comment by kevin85421** [comment]
> Good point. Indeed, we do not depend on .status.Suspend field in https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/raycluster/raycluster_controller.go

Oh, I do think that Kueue depends on the RayCluster CR status, but I just realized that it depends on `spec` instead.

```
func (j *RayCluster) IsSuspended() bool {
	return j.Spec.Suspend != nil && *j.Spec.Suspend
}
```

It is possible for this function to return true if the RayCluster is in the process of suspending. Will this be an issue?

**Comment by tenzen-y** [comment]
> > Good point. Indeed, we do not depend on .status.Suspend field in https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/raycluster/raycluster_controller.go
> 
> Oh, I do think that Kueue depends on the RayCluster CR status, but I just realized that it depends on `spec` instead.
> 
> ```
> func (j *RayCluster) IsSuspended() bool {
> 	return j.Spec.Suspend != nil && *j.Spec.Suspend
> }
> ```
> 
> It is possible for this function to return true if the RayCluster is in the process of suspending. Will this be an issue?

IIUC, GenericJob framework interface `IsSuspsnded` function indicates if desired object is suspended.
So, the current RayCluster integration seems like better. Do you have any view that Kueue should observe the interium `Suspending` condition?

When does RayCluster have `Suspending` condition? Is it after the specified .spec.suspend=true?
